### PR TITLE
Jetpack Onboarding: Track title and description change

### PR DESCRIPTION
**Pull request by tyxla**
This PR updates the site title step to track when any of the site title and site description fields are changed. We don't need to track every single change (adding or removing a character) - that would result in a ton of events anyway; we only need to know whether the user typed in a field at some point (at least once). That is why this PR only tracks the first change of each of the fields; we're not interested in any subsequent changes.

Note: we're using a non-static property because this way we'll track the change event every time 
the site title step is loaded, and not just the first time (that matters only if the user goes back to the site title step, of course). 

This PR is part of #21686.

To test:
* Checkout this branch
* Start the onboarding flow for a Jetpack site by going to `/wp-admin/admin.php?page=jetpack&action=onboard&calypso_env=development`
* Enable debug of tracks in Calypso by typing this in your browser console: `localStorage.setItem('debug', 'calypso:analytics:tracks' )`
* Input something in the site title field and observe the console.
* Verify you can see the `calypso_jpo_blogname_changed` event being queued for recording.
* Verify the event is only recorded only once, and if you continue typing/deleting, it's no longer recorded.
* Input something in the site description field and observe the console.
* Verify you can see the `calypso_jpo_blogdescription_changed` event being queued for recording.
* Verify the event is only recorded only once, and if you continue typing/deleting, it's no longer recorded.
* Go to the next step.
* Go back to the site title step.
* Type something in each of the fields, and verify the change event is being recorded again, but just once.

### CODE REVIEW

File: client/jetpack-onboarding/steps/site-title.jsx [hunk 1]
```diff
@@ -37,7 +42,16 @@ class JetpackOnboardingSiteTitleStep extends React.PureComponent {
 		}
 	}
 
+	maybeTrackChangedField( fieldName, newValue ) {
+		if ( ! this.changed[ fieldName ] && newValue !== this.state[ fieldName ] ) {
+			this.props.recordJpoEvent( `calypso_jpo_${ fieldName }_changed` );
```
Comment (ockham) [inline client/jetpack-onboarding/steps/site-title.jsx]: No programmatically generated event names pls 😬

**Comment by ockham** [comment]
Reading Jesse's request that triggered this PR,

> Did they fill out a title or leave it? 
> Did they fill out a description or leave it?

my gut reaction would've been to add tracking to the `handleSubmit` handler, and compare `this.state.blogname` and `.description` to their 'previous state' (i.e. `this.props.settings.siteTitle` and  `.siteDescription`, respectively). This would avoid the need for the `changed` var, which I think would be cleaner. Curious to hear your opinion. :slightly_smiling_face:

**Comment by ockham** [comment]
If you go with my suggestion, you might actually consider dropping the `calypso_jpo_*_changed` event altogether, and add a `changed` or `modified` (or whatever) bool attr to #22383's `calypso_jpo_*_submitted` events instead.

**Comment by tyxla** [comment]
> my gut reaction would've been to add tracking to the `handleSubmit` handler, and compare `this.state.blogname` and `.description` to their (pseudo-initial) 'previously saved state' (i.e. `this.props.settings.siteTitle` and  `.siteDescription`, respectively). This would avoid the need for the `changed` var, which I think would be cleaner. Curious to hear your opinion. :slightly_smiling_face: 

Well, that's the thing - if we go with what you suggest, how do we know if someone actually filled a title and failed to submit the step if they didn't submit but decided to skip? Also, I think having separate `change` and `submit` events would be helpful if we decide to analyze in-depth the behavior of users on a particular step.

> On a practical level, it is hard for me to imagine what insights such granular info would provide re product design

It might be the case we're overly granular here, I agree. But IMHO if I get to analyze in-depth how everybody behaves on a certain step, it would be useful to have some granularity and differentiation between the `change` and `submit` events. One might submit without changing; and one might change without submitting, so these are pretty granular actions in my head. Also: would it hurt to know more?

I'm not too heavy on my opinion, I just think that having alternative options to understand and analyze the data is always helpful.

Let me know what you folks think.

File: client/jetpack-onboarding/steps/site-title.jsx (hunk 1 above)
Comment (tyxla) [reply client/jetpack-onboarding/steps/site-title.jsx]: Good catch, fixed in https://github.com/Automattic/wp-calypso/pull/22379/commits/c5f83f96a92384781d00607135a170243c7000e1

**Comment by ockham** [comment]
> Well, that's the thing - if we go with what you suggest, how do we know if someone actually filled a title and failed to submit the step if they didn't submit but decided to skip?

> One might submit without changing; and one might change without submitting, so these are pretty granular actions in my head.

In terms of product design, I was mostly reading the request for this kind of tracking as a means to find out whether users were actually using what we're offering, or just leaving the fields unchanged and hitting the CTA -- which isn't the same as using the 'Skip to next step' button, but effectively means they're skipping what this step offers them.

Granted, that's an assumption on my side. However, weighing the need for a new instance var (`changed`)  vs just using what we have (while losing one bit of information -- did they change the fields, but instead of pressing the CTA, decided to skip after all?), I'd be okay with the latter, since that bit of information seems almost negligible to me, TBH, and I'd rather avoid the added complexity (however small) that comes with the `changed` var.

**Comment by tyxla** [comment]
@ockham what you suggest implies that we would be able to differentiate between the following cases:

* The field has been changed and the form was submitted
* The field has not been changed and the form was submitted

but we would not differentiate between the following cases:

* The field has been changed and the form was not submitted or was skipped
* The field has not been changed and the form was not submitted or was skipped

To me, the "fill out a title" would apply to the latter as well, which is why I went with this approach. If we don't care about differentiating between the latter, I'm fine with the approach you suggest, or even with what @AnnaMag suggests (not tracking `changed` at all, and closing the PRs for that).

**Comment by ockham** [comment]
Jesse's list included

> Did they skip? 
> Did they fill out a title or leave it? 
> Did they fill out a description or leave it?
> [...]
> Complete the step [...]

which seems to indicate he wants information whether the user changed either of the input fields tracked separately (i.e. not just skipping/submission of the form). Which is why I didn't consider dropping it altogether :slightly_smiling_face:

**Comment by tyxla** [comment]
Well, that's the thing - if we hook tracking to the submission handler, we won't be tracking when the user change and decide to skip or just not submit. Thus my approach 😉

**Comment by ockham** [comment]
Hmm, I might've misread you then. I thought it was about closing this PR altogether, thus not tracking changes at all.

I'll restate my reading of things. and my conclusions:

We want to track if the user changes title/description, or if they leave them be.

1. Leave them be can mean clicking 'Skip this step', which AFAIK we track. I don't think we need any more information in this case, i.e. whether they ever changed the text fields before 'cancelling' this step. In my reading, the latter is an edge case with little relevant information.
2. It can also mean not changing the fields, but clicking the CTA. We can track this in the submit handler, adding an additional property to the event, and comparing `state` to `props.settings`. Given the size and appearance of the button compared to the 'Skip' link, this is a far more realistic scenario.
3. Changing them means actually changing either field, and clicking the CTA. Can be covered the same way as 2.

**Comment by tyxla** [comment]
Makes sense, you convinced me @ockham 😃 

Feel free to have a new look now.

**Comment by ockham** [comment]
(As manifested by the code I mean.)

**Pull request merged**